src/app_model/types/_command_rule.py

```diff
@@ -7,6 +7,29 @@
 from ._icon import Icon
 
 
+class ToggleRule(_StrictModel):
```

Comment:
@Czaki, if you have a moment to take a look, would appreciate your thoughts.

This is my current strategy for getting the viewer toggle action to work in https://github.com/napari/napari/pull/4826

(see https://github.com/napari/napari/pull/4826/files#diff-b93fcae709b97acc12b9c786489465f3d8b8dcb186d5035c0df9a7b49b7b0e36 for the napari side usage)

the "better" long-term solution is to just use a when clause to determine whether the toggle should be checked or not.  but that would require a lot more work on the side of context names in napari.  This provides a more direct hookup (akin to what we're already doing in napari).  

The obvious downside is that it passes the backend object directly... hence the name `experimental_connect`.  I figure we'll have the context/when clauses finished long before we use a backend other than Qt ... so I (personally) feel like this workaround is fine for the short/medium term.  

thoughts?

Reply:
At this moment, it is not clear to me how it will work for viewer. Did condition will be passed to injection manager that will provide values to it calculate? 

Both `initialize` and `experimental_connect` are pointed to be called only on creation. So will not allow to synchronize state changed from another place.

Reply:
it's supposed to work as follows:
1. viewer.window gets created
2. menus get built, calling `build_qmodel_menu` for each of the top menus
3. `QModelMenu` then creates a `QMenuItemAction` for each action in the menu
4. during the init of the MenuItemAction, `experimental_connect()` gets called.
5. at that point, we have a current viewer, so if the callback provided to experimental connect has any annotations, they can be injected.

There _is_ however, a sneaky little bug in here that needs to be fixed; and that is that `QMenuItemAction` objects are cached (so that different representations of the same command in a menu, button, etc. all point to the same QAction)... however, that caching means that step 4/5 are only getting called for the first viewer.  So indeed, the second viewer isn't synced.  need to ponder that :/

Reply:
So determine state on open menu by evaluating conditions will solve this problem?

Reply:
ah, that's definitely worth a try!  That could completely replace the need for experimental connect huh?  It would mean that any changes on the napari side wouldn't be reflected in an "open" menu... but that's such a minor concern, and will almost never happen.  so i think I like it!  will update when I try it

Reply:
The standard method of invoking the menu https://doc.qt.io/qt-6/qmenu.html#exec should prevent any changes from the main thread.

Reply:
yeah... actually though.  I just thought of a downside of this: which is that menus are just one place where a toggleable qaction might appear.  They can also back QToolbuttons (which app-model also supports)... and in that case, when you're not opening a menu, it's less obvious when one should call "refresh".  

So, there's definitely a big benefit of using the "on menu open" approach to check toggle state, and it would mean we wouldn't have to expose the backend at all (which i love).  but it wouldn't work for qtoolbuttons: and ultimately, I think that all of our layer mode buttons should be toolbuttons.

However, that's probably even farther down the road, and at that point we could just say, "if you want a responsive toolbutton, you have to use a context condition".  Seem good?

Reply:
Hm. And what did you think about passing a tuple `(Annotation_for_injection, `path.to.signal.to.be.connected`) instead of function? this will force a single event but allow for proper connection and disconnect when creating/deleting backend objects.

Reply:
I just made a new PR for you to review what it might look like to evaluate conditions whenever a menu is being show: https://github.com/napari/app-model/pull/59

let me have a think now about this tuple strategy...